src/Arcus.Messaging.Pumps.Abstractions/MessagePump.cs

```diff
@@ -0,0 +1,124 @@
+            Guard.NotNull(configuration, nameof(configuration));
+            Guard.NotNull(serviceProvider, nameof(serviceProvider));
+
+            Logger = logger;
+            Configuration = configuration;
+            ServiceProvider = serviceProvider;
+        }
+
+        /// <summary>
+        ///     Handles an exception that occured during the receiving of a message
+        /// </summary>
+        /// <param name="receiveException">Exception that occured</param>
+        protected virtual Task HandleReceiveExceptionAsync(Exception receiveException)
+        {
+            Logger.LogError(receiveException, "Unable to process message");
+            return Task.CompletedTask;
+        }
+
+        /// <summary>
+        ///     Deserializes a raw JSON message body
+        /// </summary>
+        /// <param name="rawMessageBody">Raw message body to deserialize</param>
+        /// <param name="messageContext">Context concerning the message</param>
+        /// <returns>Deserialized message</returns>
+        protected virtual TMessage DeserializeJsonMessageBody(byte[] rawMessageBody, MessageContext messageContext)
+        {
+            Encoding encoding = DetermineMessageEncoding(messageContext);
+            string serializedMessageBody = encoding.GetString(rawMessageBody);
+
+            TMessage messageBody = JsonConvert.DeserializeObject<TMessage>(serializedMessageBody);
+            if (messageBody == null)
+            {
+                Logger.LogError("Unable to deserialize to message contract {ContractName} for message {MessageBody}",
+                    typeof(TMessage), rawMessageBody);
+            }
+
+            return messageBody;
+        }
+
+        /// <summary>
+        ///     Process a new message that was received
+        /// </summary>
+        /// <param name="message">Message that was received</param>
+        /// <param name="messageContext">Context providing more information concerning the processing</param>
+        /// <param name="correlationInfo">
+        ///     Information concerning correlation of telemetry & processes by using a variety of unique
+        ///     identifiers
+        /// </param>
+        /// <param name="cancellationToken">Cancellation token</param>
+        protected abstract Task ProcessMessageAsync(TMessage message, TMessageContext messageContext,
+            MessageCorrelationInfo correlationInfo, CancellationToken cancellationToken);
+
+        /// <summary>
+        ///     Determines the encoding used for a given message
+        /// </summary>
+        /// <remarks>If no encoding was specified, UTF-8 will be used by default</remarks>
+        /// <param name="messageContext">Context concerning the message</param>
+        /// <returns>Encoding that was used for the message body</returns>
+        protected Encoding DetermineMessageEncoding(MessageContext messageContext)
+        {
+            var encoding = Encoding.UTF8;
```

Comment:
This is nitpicking. I'm not such a fan of rewriting variables, so if I could prevent it, I'll do it.
Here, you could directly from in the `try...catch` and return the default UTF-8 afterwards. Removing the need for a temporary variable that needs to be overwritten anyway.

Reply:
```csharp
protected Encoding DetermineMessageEncoding(MessageContext messageContext)
        {
            var encoding = Encoding.UTF8;
            if (messageContext.Properties.TryGetValue(PropertyNames.Encoding, out object annotatedEncoding))
                try
                {
                    encoding = Encoding.GetEncoding(annotatedEncoding.ToString());
                }
                catch (Exception ex)
                {
                    Logger.LogCritical(ex,
                        $"Unable to determine encoding with name '{{Encoding}}'. Falling back to {{FallbackEncoding}}.",
                        annotatedEncoding.ToString(), encoding.WebName);
                }

            return encoding;
        }
```

Could become this:
```csharp
            if (messageContext.Properties.TryGetValue(PropertyNames.Encoding, out object annotatedEncoding))
            {
                try
                {
                    return Encoding.GetEncoding(annotatedEncoding.ToString());
                }
                catch (Exception ex)
                {
                    Logger.LogCritical(ex,
                        $"Unable to determine encoding with name '{{Encoding}}'. Falling back to {{FallbackEncoding}}.",
                        annotatedEncoding.ToString(), Encoding.UTF8.WebName);
                }
            }

            return Encoding.UTF8;
```

But then we duplicate the default encoding twice.

That's fine for me, but i'd suggest to introduce a global const in the class with the default encoding then, ok for you?

Reply:
I'll leave it up to you. It's not a very critical place, so both approaches are good for me.

Reply:
Changed it